src/redBlackTree.jl

```diff
@@ -0,0 +1,111 @@
+import Base: haskey, getindex, setindex!
+
+export Tree, EmptyTree, RBNode, RedBlackTree
+
+abstract Tree{K,V}
+
+type EmptyTree{K,V} <: Tree{K,V}
+end
+
+type RBNode{K,V} <: Tree{K,V}
+    key :: K
+    data :: V
+    left :: Tree{K,V}
+    right :: Tree{K,V}
```

Comment:
This is probably the most straightforward implementation.  Performance-wise, however, in Julia, it's a [bad idea to have abstract members of types](http://docs.julialang.org/en/release-0.4/manual/performance-tips/#avoid-fields-with-abstract-type).

A workaround: https://github.com/JuliaLang/julia/issues/15791#issuecomment-207060497

(You'll probably have to read the whole thread for context.)